### PR TITLE
library

### PR DESCRIPTION
**Pull request by drozddmitriy**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+source 'https://rubygems.org'
+
+gem 'faker', '~> 1.9', '>= 1.9.3'
```
Comment (dzemlianoi-double) [inline Gemfile]: ffaker

File: core/library.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
+require './dependency'
```
Comment (dzemlianoi-double) [inline core/library.rb]: let's require in entry file

File: core/library.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+class Library
+  include Database
+  include Generator
+
+  attr_accessor :masauthor, :masbook, :masorder, :masreader
```
Comment (dzemlianoi-double) [inline core/library.rb]: authors, books, orders, readers

File: core/library.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+  def initialize
+    load_entities
+  end
+
+  def create_book(title, auth)
```
Comment (dzemlianoi-double) [inline core/library.rb]: separate method `add`

Comment (dzemlianoi-double) [inline core/library.rb]: author

File: core/library.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def create_reader(name, email, city, street, house)
+    raise Myvalid, 'house should be a Integer and positive number!' unless (house.is_a? Integer) && house.positive?
+
+    @masreader.push(Reader.new(name, email, city, street, house))
+  end
+
+  def create_order(book, reader, date = Date.today)
+    raise Myvalid, 'wrong argument book or reader!' unless (book.is_a? Book) && (reader.is_a? Reader)
```
Comment (dzemlianoi-double) [inline core/library.rb]: separate methods

File: core/library.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    @masorder.push(Order.new(book, reader, date))
+  end
+
+  def top_reader(quantity = 1)
```
Comment (dzemlianoi-double) [inline core/library.rb]: separate module

File: core/moduledb.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+module Database
+  def load_entities
+    raise Myvalid, 'file not found!' unless File.exist?('data.yml')
```
Comment (dzemlianoi-double) [inline core/moduledb.rb]: 1) Myvalid
2) magic strings

File: core/moduledb.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+    data = YAML.load_file 'data.yml'
```
Comment (dzemlianoi-double) [inline core/moduledb.rb]: data = YAML.load_file 'data.yml' || {}

File: core/moduledb.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    if data == false
+      @masauthor = []
+      @masbook = []
+      @masorder = []
+      @masreader = []
+    else
+      @masauthor = data[:masauthor]
```
Comment (dzemlianoi-double) [inline core/moduledb.rb]: data[:masauthor] || []

File: core/library.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+  def most_popular_books(quantity = 1)
+    hash = @masorder.each_with_object(Hash.new(0)) { |v, h| h[v.book.title] += 1; }
+
+    raise Myvalid, 'quantity is faile!' unless hash.count >= quantity
+
+    hash.keys.reverse_each.first(quantity)
+  end
+
+  def num_of_readers_of_most_popular_books(quantity = 3)
```
Comment (dzemlianoi-double) [inline core/library.rb]: @MaximKoptsov, here!!1

File: core/moduledb.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      @masbook = data[:masbook]
+      @masorder = data[:masorder]
+      @masreader = data[:masreader]
+    end
+  end
+
+  def save
```
Comment (dzemlianoi-double) [inline core/moduledb.rb]: maybe to_h

File: entities/author.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+class Author
+  attr_accessor :name, :biography
```
Comment (dzemlianoi-double) [inline entities/author.rb]: attr_accessor?

File: entities/book.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+class Book
+  attr_accessor :title, :auth
+
+  def initialize(title, auth)
+    @title = title
+    @auth = auth
```
Comment (dzemlianoi-double) [inline entities/book.rb]: author?

File: entities/reader.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+class Reader
+  attr_accessor :name, :email, :city, :street, :house
+
+  def initialize(name, email, city, street, house)
+    @name = name
+    @email = email
+    @city = city
+    @street = street
+    @house = Integer house
```
Comment (dzemlianoi-double) [inline entities/reader.rb]: to_i

File: Gemfile (hunk 1 above)
Comment (za4emnik) [inline Gemfile]: change to ffaker, please

File: core/modulevalidation.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+module ValidationEntity
```
Comment (za4emnik) [inline core/modulevalidation.rb]: use snake_case for a file name

Comment (za4emnik) [reply core/modulevalidation.rb]: check and change everywhere

File: core/validation.rb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+class Validation < StandardError
```
Comment (za4emnik) [inline core/validation.rb]: naming is not clear
try to rename, e.g. DataValidation or smth else

File: main.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
+require './dependency'
+
+generator = Generator.new
+library = Library.new
+
+################generate Library##################################
+# authors = generator.generate_authors
```
Comment (za4emnik) [inline main.rb]: why comments?

File: core/library.rb [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    else raise Validation, 'entity is faile!'
+    end
+  end
+
+  def top_reader(quantity = 1)
+    hash = @orders.each_with_object(Hash.new(0)) { |v, h| h[v.reader.name] += 1; }
+
+    raise Validation, 'quantity is faile!' unless hash.count >= quantity
+
+    hash.keys.reverse_each.first(quantity)
```
Comment (za4emnik) [inline core/library.rb]: you simply calculate the sum of readers, bt you need return top readers

File: core/library.rb [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  def most_popular_books(quantity = 1)
+    hash = @orders.each_with_object(Hash.new(0)) { |v, h| h[v.book.title] += 1; }
+
+    raise Validation, 'quantity is faile!' unless hash.count >= quantity
+
+    hash.keys.reverse_each.first(quantity)
```
Comment (za4emnik) [inline core/library.rb]: the same

**Review by za4emnik — CHANGES_REQUESTED** [review]
1) Fix review comments
2) Check statistics logic
3) Try to create `Statistics` module with statistics and include it to the library
4) Fix rubocop offenses
5) Commit messages should start with a capital letter and more informative. For example `delete module` - bad commit, `Delete module MyModule` - good commit

File: core/module_db.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
+module Database
```
Comment (za4emnik) [inline core/module_db.rb]: name of file `module_db.rb` does not match the name of module `module Database`

File: core/module_statistic.rb [hunk 21]
```diff
@@ -0,0 +1,19 @@
+module Statistic
```
Comment (za4emnik) [inline core/module_statistic.rb]: the same

File: core/module_validation.rb [hunk 22]
```diff
@@ -0,0 +1,9 @@
+module ValidationEntity
```
Comment (za4emnik) [inline core/module_validation.rb]: check all files, please

File: core/module_statistic.rb [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+    orders.group_by(&:reader).sort_by { |_key, order| -order.count }.first(quantity)
+  end
+
+  def most_popular_books(orders, quantity)
+    orders.group_by(&:book).sort_by { |_key, order| -order.count }.first(quantity)
+  end
+
+  def num_of_readers_of_most_popular_books(orders, quantity)
+    array_books = []
+    most_books = most_popular_books(orders, quantity)
```
Comment (za4emnik) [inline core/module_statistic.rb]: Why so hard? We can just smth like this: `most_books.to_h.keys.map(&:title)`

**Review by za4emnik — CHANGES_REQUESTED** [review]
1) Check name of files
2) Fix rubocop offenses

File: core/validation.rb [hunk 24]
```diff
@@ -0,0 +1,2 @@
+class BookValidation < StandardError
```
Comment (za4emnik) [inline core/validation.rb:1]: remove

**Review by za4emnik — APPROVED** [review]
Merge after fix

**Pull request merged**